dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -2501,58 +2786,115 @@ RtpsUdpDataLink::send_heartbeats()
+  RtpsUdpInst& cfg = link->config();
+
+  typedef ReaderInfoMap::iterator ri_iter;
+  const ri_iter end = remote_readers_.end();
+  for (ri_iter ri = remote_readers_.begin(); ri != end; ++ri) {
+    if (has_data || !ri->second.handshake_done_) {
+      response.to_guids_.insert(ri->first);
+      if (is_final && !ri->second.handshake_done_) {
+        is_final = false;
+      }
+    }
+    if (!ri->second.durable_data_.empty()) {
+      const ACE_Time_Value expiration =
+        ri->second.durable_timestamp_ + cfg.durable_data_timeout_;
+      if (now > expiration) {
+        typedef OPENDDS_MAP(SequenceNumber, TransportQueueElement*)::iterator
+          dd_iter;
+        for (dd_iter it = ri->second.durable_data_.begin();
+             it != ri->second.durable_data_.end(); ++it) {
+          pendingCallbacks.push_back(it->second);
+        }
+        ri->second.durable_data_.clear();
+        if (Transport_debug_level > 3) {
+          const GuidConverter gw(id_), gr(ri->first);
+          VDBG_LVL((LM_INFO, "(%P|%t) RtpsUdpDataLink::send_heartbeats - "
+            "removed expired durable data for %C -> %C\n",
+            OPENDDS_STRING(gw).c_str(), OPENDDS_STRING(gr).c_str()), 3);
+        }
+      } else {
+        has_durable_data = true;
+        if (ri->second.durable_data_.rbegin()->first > durable_max) {
+          durable_max = ri->second.durable_data_.rbegin()->first;
         }
+        response.to_guids_.insert(ri->first);
       }
-      std::iter_swap(it, --last);
-    } else {
-      ++it;
     }
   }
-  hbs.erase(last, hbs.end());
-#else
-  ACE_UNUSED_ARG(hbs);
+
+  if (!elems_not_acked_.empty()) {
+    is_final = false;
+  }
+
+  if (is_final && !has_data && !has_durable_data) {
+    return true;
+  }
+
+  if (!ready_to_hb_) {
+    return true;
+  }
+
+  const SequenceNumber firstSN = (durable_ || !has_data) ? 1 : send_buff_->low(),
+    lastSN = std::max(durable_max, has_data ? send_buff_->high() : SequenceNumber::ZERO());
+
+  using namespace OpenDDS::RTPS;
+
+  const HeartBeatSubmessage hb = {
+    {HEARTBEAT,
+     CORBA::Octet(FLAG_E | (is_final ? FLAG_F : 0)),
+     HEARTBEAT_SZ},
+    ENTITYID_UNKNOWN, // any matched reader may be interested in this
+    id_.entityId,
+    {firstSN.getHigh(), firstSN.getLow()},
+    {lastSN.getHigh(), lastSN.getLow()},
+    {++heartbeat_count_}
+  };
+  response.sm_.heartbeat_sm(hb);
+
+#ifdef OPENDDS_SECURITY
+  const EntityId_t& volatile_writer =
+    RTPS::ENTITYID_P2P_BUILTIN_PARTICIPANT_VOLATILE_SECURE_WRITER;
+  if (std::memcmp(&id_.entityId, &volatile_writer, sizeof(EntityId_t)) == 0) {
+    RepoIdSet guids = response.to_guids_;
+    response.dst_guid_ = GUID_UNKNOWN;
+    for (RepoIdSet::const_iterator it = guids.begin(); it != guids.end(); ++it) {
+      response.to_guids_.clear();
+      response.dst_guid_ = (*it);
+      response.sm_.heartbeat_sm().readerId = it->entityId;
+      responses.push_back(response);
+    }
+  } else
```

Comment:
May be easier to `return` here instead of splitting the if/else across the conditional block

Reply:
I think it's OK to just duplicate this one line in the #else portion for code clarity.